File: WinFormsClient/PacketDefinition.cs
using System.Text;

namespace WinFormsClient;

internal class PacketDefinition
{
    public enum PacketID : ushort
    {
        ReqLogin = 1002,
        ResLogin = 1003,
        ReqRoomEnter = 1021,
        ResRoomEnter = 1022,
        ReqRoomChat = 1026,
        ResRoomChat = 1027,
    }

    public class PacketHeader
    {
        public ushort TotalSize { get; set; }
        public PacketID Id { get; set; }
        public byte Type { get; set; }

        public byte[] Serialize()
        {
            byte[] buffer = new byte[6];
            BitConverter.GetBytes(TotalSize).CopyTo(buffer, 0);
            BitConverter.GetBytes((ushort)Id).CopyTo(buffer, 2);
            buffer[4] = Type;
            return buffer;
        }

        public static PacketHeader Deserialize(byte[] buffer)
        {
            return new PacketHeader
            {
                TotalSize = BitConverter.ToUInt16(buffer, 0),
                Id = (PacketID)BitConverter.ToUInt16(buffer, 2),
                Type = buffer[4]
            };
        }
    }

    public class LoginRequest : PacketHeader
    {
        public string UserID { get; set; }
        public string AuthToken { get; set; }

        public byte[] Serialize()
        {
            byte[] headerBuffer = base.Serialize();
            byte[] userIdBuffer = Encoding.UTF8.GetBytes(UserID.PadRight(32, '\0'));
            byte[] authTokenBuffer = Encoding.UTF8.GetBytes(AuthToken.PadRight(32, '\0'));

            byte[] buffer = new byte[headerBuffer.Length + userIdBuffer.Length + authTokenBuffer.Length];
            Array.Copy(headerBuffer, 0, buffer, 0, headerBuffer.Length);
            Array.Copy(userIdBuffer, 0, buffer, headerBuffer.Length, userIdBuffer.Length);
            Array.Copy(authTokenBuffer, 0, buffer, headerBuffer.Length + userIdBuffer.Length, authTokenBuffer.Length);

            return buffer;
        }

        public static LoginRequest Deserialize(byte[] buffer)
        {
            return new LoginRequest
            {
                TotalSize = BitConverter.ToUInt16(buffer, 0),
                Id = (PacketID)BitConverter.ToUInt16(buffer, 2),
                Type = buffer[4],
                UserID = Encoding.UTF8.GetString(buffer, 6, 32).TrimEnd('\0'),
                AuthToken = Encoding.UTF8.GetString(buffer, 38, 32).TrimEnd('\0')
            };
        }
    }

    public class RoomEnterRequest : PacketHeader
    {
        public int RoomNumber { get; set; }

        public byte[] Serialize()
        {
            byte[] headerBuffer = base.Serialize();
            byte[] roomNumberBuffer = BitConverter.GetBytes(RoomNumber);

            byte[] buffer = new byte[headerBuffer.Length + roomNumberBuffer.Length];
            Array.Copy(headerBuffer, 0, buffer, 0, headerBuffer.Length);
            Array.Copy(roomNumberBuffer, 0, buffer, headerBuffer.Length, roomNumberBuffer.Length);

            return buffer;
        }

        public static RoomEnterRequest Deserialize(byte[] buffer)
        {
            return new RoomEnterRequest
            {
                TotalSize = BitConverter.ToUInt16(buffer, 0),
                Id = (PacketID)BitConverter.ToUInt16(buffer, 2),
                Type = buffer[4],
                RoomNumber = BitConverter.ToInt32(buffer, 6)
            };
        }
    }

    public class RoomChatRequest : PacketHeader
    {
        public string Message { get; set; }

        public byte[] Serialize()
        {
            byte[] headerBuffer = base.Serialize();
            byte[] messageBuffer = Encoding.UTF8.GetBytes(Message);

            byte[] buffer = new byte[headerBuffer.Length + messageBuffer.Length];
            Array.Copy(headerBuffer, 0, buffer, 0, headerBuffer.Length);
            Array.Copy(messageBuffer, 0, buffer, headerBuffer.Length, messageBuffer.Length);

            return buffer;
        }

        public static RoomChatRequest Deserialize(byte[] buffer)
        {
            return new RoomChatRequest
            {
                TotalSize = BitConverter.ToUInt16(buffer, 0),
                Id = (PacketID)BitConverter.ToUInt16(buffer, 2),
                Type = buffer[4],
                Message = Encoding.UTF8.GetString(buffer, 6, buffer.Length - 6)
            };
        }
    }
}

File: ServerAcl/PacketDefinition.cpp
#include "PacketDefinition.h"

void PacketHeader::Serialize(char* buffer) const {
    std::memcpy(buffer, &TotalSize, sizeof(TotalSize));
    std::memcpy(buffer + sizeof(TotalSize), &Id, sizeof(Id));
    std::memcpy(buffer + sizeof(TotalSize) + sizeof(Id), &Type, sizeof(Type));
}

PacketHeader PacketHeader::Deserialize(const char* buffer) {
    PacketHeader header;
    std::memcpy(&header.TotalSize, buffer, sizeof(header.TotalSize));
    std::memcpy(&header.Id, buffer + sizeof(header.TotalSize), sizeof(header.Id));
    std::memcpy(&header.Type, buffer + sizeof(header.TotalSize) + sizeof(header.Id), sizeof(header.Type));
    return header;
}

void LoginRequest::Serialize(char* buffer) const {
    PacketHeader::Serialize(buffer);
    std::memcpy(buffer + sizeof(PacketHeader), UserID, sizeof(UserID));
    std::memcpy(buffer + sizeof(PacketHeader) + sizeof(UserID), AuthToken, sizeof(AuthToken));
}

LoginRequest LoginRequest::Deserialize(const char* buffer) {
    LoginRequest request;
    request.TotalSize = *(uint16_t*)buffer;
    request.Id = *(PacketID*)(buffer + sizeof(request.TotalSize));
    request.Type = buffer[sizeof(request.TotalSize) + sizeof(request.Id)];
    std::memcpy(request.UserID, buffer + sizeof(PacketHeader), sizeof(request.UserID));
    std::memcpy(request.AuthToken, buffer + sizeof(PacketHeader) + sizeof(request.UserID), sizeof(request.AuthToken));
    return request;
}

// RoomEnterRequest
void RoomEnterRequest::Serialize(char* buffer) const {
    PacketHeader::Serialize(buffer);
    std::memcpy(buffer + sizeof(PacketHeader), &RoomNumber, sizeof(RoomNumber));
}

RoomEnterRequest RoomEnterRequest::Deserialize(const char* buffer) {
    RoomEnterRequest request;
    request.TotalSize = *(uint16_t*)buffer;
    request.Id = *(PacketID*)(buffer + sizeof(request.TotalSize));
    request.Type = buffer[sizeof(request.TotalSize) + sizeof(request.Id)];
    std::memcpy(&request.RoomNumber, buffer + sizeof(PacketHeader), sizeof(request.RoomNumber));
    return request;
}

// RoomChatRequest
void RoomChatRequest::Serialize(char* buffer) const {
    PacketHeader::Serialize(buffer);
    std::memcpy(buffer + sizeof(PacketHeader), Message, sizeof(Message));
}

RoomChatRequest RoomChatRequest::Deserialize(const char* buffer) {
    RoomChatRequest request;
    request.TotalSize = *(uint16_t*)buffer;
    request.Id = *(PacketID*)(buffer + sizeof(request.TotalSize));
    request.Type = buffer[sizeof(request.TotalSize) + sizeof(request.Id)];
    std::memcpy(request.Message, buffer + sizeof(PacketHeader), sizeof(request.Message));
    return request;
}

// RoomChatNotification
void RoomChatNotification::Serialize(char* buffer) const {
    PacketHeader::Serialize(buffer);
    std::memcpy(buffer + sizeof(PacketHeader), UserID, sizeof(UserID));
    std::memcpy(buffer + sizeof(PacketHeader) + sizeof(UserID), Message, sizeof(Message));
}

RoomChatNotification RoomChatNotification::Deserialize(const char* buffer) {
    RoomChatNotification notification;
    notification.TotalSize = *(uint16_t*)buffer;
    notification.Id = *(PacketID*)(buffer + sizeof(notification.TotalSize));
    notification.Type = buffer[sizeof(notification.TotalSize) + sizeof(notification.Id)];
    std::memcpy(notification.UserID, buffer + sizeof(PacketHeader), sizeof(notification.UserID));
    std::memcpy(notification.Message, buffer + sizeof(PacketHeader) + sizeof(notification.UserID), sizeof(notification.Message));
    return notification;
}